src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,319 @@
+    ]
+    res_objs = list(itertools.chain.from_iterable(objs))
+    if filter_objects_name_by:
+        res_objs = [
+            obj for obj in res_objs if obj[2].startswith(filter_objects_name_by)
+        ]
+    return res_objs
+
+
+def process_input(conf):
+    """
+    Download .fa and .meta files from workspace and save a copy under output_dir
+    """
+    while True:
+        upa, obj_info = conf.queue.get(block=True)
+        if not upa:
+            print("Stopping")
+            break
+        cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+        # upa file is downloaded to cfn
+        conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+        # each upa in output_dir as a seperate directory
+        dstd = os.path.join(conf.pth, upa)
+        os.makedirs(dstd, exist_ok=True)
+
+        dst = os.path.join(dstd, f"{upa}.fa")
+        # copy downloaded upa file to output_dir. Hard link might cause invalid cross-device link problem
+        shutil.copy(cfn, dst)
+
+        metafile = os.path.join(dstd, f"{upa}.meta")
+        # save meta file with relevant object_info
+        json.dump(_process_object_info(obj_info), open(metafile, "w"), indent=2)
+
+        print("Completed %s" % (upa))
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="PROTOTYPE - Download genome files from the workspace service (WSS)."
+    )
+
+    required = parser.add_argument_group("required named arguments")
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+    required.add_argument(
+        "--project_dir",
+        type=str,
+        help="Path points to Collections repository",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--output_dir",
+        type=str,
+        help="Output directoy to save genome objects.",
+    )
+    optional.add_argument(
+        "--job_dir",
+        type=str,
+        help="Job directoy of source link",
+    )
+    optional.add_argument(
+        "--source", type=str, default="WS", help="Source of data (default: WS)"
+    )
+    optional.add_argument(
+        "--workers",
+        type=int,
+        default=5,
+        help="Number of workers for multiprocessing",
+    )
+    optional.add_argument(
+        "--overwrite", action="store_true", help="Overwrite existing files."
+    )
+    optional.add_argument(
+        "--batch_size", type=int, default=10000, help="Batch size of object id"
+    )
+
+    args = parser.parse_args()
+
+    (
+        workspace_id,
+        project_dir,
+        root_dir,
+        output_dir,
+        job_dir,
+        source,
+        workers,
```

Comment:
do we need `workers = max(args.workers, 1)`? or better do it in `add_argument`.

Reply:
add custom type function for positive val.

Reply:
This seems like a less verbose way to do the same thing, since argparse takes care of the int coercion / checking:
https://stackoverflow.com/questions/18700634/python-argparse-integer-condition-12/35083753#35083753

Reply:
Updated with parser.error.

Reply:
Updated with choice = range(1, cpu_count() + 1)

Reply:
The problem with using choice is if there's 200 processors the help message will include a list like `1, 2, ..., 200`. That's why I suggested the solution given in the link above

Reply:
Used the solution given in the link above.

Reply:
:+1: